`Engine::~Engine()` should wait for non-reentrant threads to shutdown

malfet opened the pull request:
Because `this` must be valid while `Engine::main_thread` is running, at least for non-reentrant worker threads

Test Plan: Run `test_api --gtest-filter=ModulesTest.InstanceNorm1d` in a loop

albanD reviewed (commented):
Thanks for the PR.
So the problem is just that we finish destroying the engine before the shutdown tasks are processed?
Not that there is still a backward running and the leaked threads won't have access to the engine?

EscapeZero commented:
@albanD `this` is being passed into a child thread that is being `detached` at which point you have no way of ensuring the thread is finished with `this` before it is destructed.

The OS could theoretically never schedule the detached threads before `Engine` is destructed in its current state.

albanD reviewed (approved):
Sounds good.
Thanks !

malfet commented:
> So the problem is just that we finish destroying the engine before the shutdown tasks are processed?
Yes, there is no guarantee about the order and I can reproduce the failure pretty reliably by running multiple instances of the test concurrently.
> Not that there is still a backward running and the leaked threads won't have access to the engine?
Is there a reason why I shouldn't wait for those as well?

@albanD , any objection if I explicitly delete  `Engine` class copy constructor?

albanD reviewed (changes requested):
@malfet  Given that we should only ever have a single instance of the Engine. It sounds ok.

Given the CI errors, I guess it deadlocks on deletion?
This should be fixed.

malfet commented:
@albanD modified `Engine` destructor to simply wait for non-reentrant thread to decrease atomic counter, which still achieves the same goal but solves the deadlock in `PythonEngine`, when worker thread can not be joined while native code is running.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -931,6 +935,7 @@ auto Engine::start_threads() -> void {
   thread_pool_shared_ = std::make_shared<ThreadPoolShared>();
 
   for (int i = 0; i < num_threads; ++i) {
+    non_reentrant_thread_count_.fetch_add(1);
```
You don't really need an atomic additions for each one here. You can just set it to `num_threads` above.
This function is not expected to be thread safe as it only ever called inside `std::call_once()`.

malfet replied:
Updated

colesbury reviewed (commented):
Thanks for catching and fixing this. A few concerns:

1) How does this interact with fork (i.e. multiprocessing)? We have code that explicitly calls the engine destructor after fork. At that point all the threads are dead, so they cannot decrement the counter:

https://github.com/pytorch/pytorch/blob/6f8a8e4e47a37251ba82101ccc684b8bd7027b38/torch/csrc/autograd/python_engine.cpp#L96-L108

2) I would prefer we avoid active spinning. We can either remember the threads and call `thread.join()` or use a `condition_variable` + `mutex` with `non_reentrant_thread_count_`.

3) What happens in the case where a backwards is running (i.e. `noBackward` is false)? @albanD  says we're not handling that case in this PR. That's fine -- at least this is addressing the common case, but that case seems particularly susceptible to this bug.

EscapeZero commented:
Another random concern. How did our internal version of the unit tests in ovrsource catch this, but none of the fbcode or CircleCI tests catch this?

malfet commented:
@colesbury thank you for the detailed, review. Comments are below.
> 1. How does this interact with fork (i.e. multiprocessing)? We have code that explicitly calls the engine destructor after fork. At that point all the threads are dead, so they cannot decrement the counter:
Updated the commit by adding `release_workers()` method, that is called at from `_maybe_reinitialize_engine_after_fork`
> 1. I would prefer we avoid active spinning. We can either remember the threads and call `thread.join()` or use a `condition_variable` + `mutex` with `non_reentrant_thread_count_`.
Initial version of the change used `std::thread::join()`, but that caused a deadlock with python runtime. 
> 2. What happens in the case where a backwards is running (i.e. `noBackward` is false)? @albanD  says we're not handling that case in this PR. That's fine -- at least this is addressing the common case, but that case seems particularly susceptible to this bug.
Plan to address that as well, but perhaps in a separate commit, since it requires a much more rigorous testing.

albanD reviewed (commented):
I think there is a confusion in the latest version of the PR: the thread pool in `thread_pool_shared_` does not contain the worker threads.
In particular, this pool should remain empty as long as we don't need it.

Also since this datastructure is unrelated to the actual worker threads that we try to shutdown, this might not be the best place to put the counter.

malfet commented:
Ok, moved the `non_reentrant_thread_count`, condvar and mutex to private methods of `Engine` class

albanD commented on torch/csrc/autograd/engine.h:
```diff
@@ -266,6 +269,12 @@ struct TORCH_API Engine {
 
 private:
+  // Number of non-reentrant threads
+  std::atomic<uint32_t> non_reentrant_thread_count_;
```
nit: this does not need to be atomatic as all the changes are either in std::cal_once or protected by the lock below.

albanD reviewed (approved):
Looks good!

malfet committed:
`Engine::~Engine()` should wait for non-reentrant threads to shutdown

Because `this` must be valid while `Engine::main_thread` is running, at least for non-reentrant worker threads

Test Plan: Run `test_api --gtest-filter=ModulesTest.InstanceNorm1d` in a loop

malfet commented:
Finally figured out what is causing a deadlock on Windows:  CRT destroys all DLL created threads before calling global destruction. I.e. wait logic should be skipped if PyTorch was compiled as shared library on Windows.
